offload_data/plugin.cpp

```diff
@@ -0,0 +1,184 @@
+#include "common/plugin.hpp"
+#include "common/threadloop.hpp"
+#include "common/switchboard.hpp"
+#include "common/phonebook.hpp"
+#include "common/data_format.hpp"
+#include "common/global_module_defs.hpp"
+
+#include <iomanip>
+#include <fstream>
+#include <numeric>
+#define STB_IMAGE_WRITE_IMPLEMENTATION
+#include "common/gl_util/lib/stb_image_write.h"
+
+using namespace ILLIXR;
+
+
+class offload_data : public threadloop {
+public:
+	offload_data(std::string name_, phonebook* pb_)
+		: threadloop{name_, pb_}
+		, sb{pb->lookup_impl<switchboard>()}
+		, _offload_data_reader{sb->subscribe_latest<texture_pose>("texture_pose")}
+		, _seq_expect(1)
+		, _stat_processed(0)
+		, _stat_missed(0)
+		, percent{0}
+		, img_idx{0}
+		, enable_offload{ILLIXR::str_to_bool(ILLIXR::getenv_or("ILLIXR_OFFLOAD_ENABLE", "False"))}
+		, is_success{true}
+		/// TODO: Set with #198
+		, obj_dir{ILLIXR::getenv_or("ILLIXR_OFFLOAD_PATH", "metrics/offloaded_data/")}
+		{
+			/// Remove existing file and create folder for offloading
+			/// TODO: Not portable
+			system(("rm -r " + obj_dir).c_str());
+			system(("mkdir -p " + obj_dir).c_str());
+		}
+
+		virtual skip_option _p_should_skip() override {
+			auto in = _offload_data_reader->get_latest_ro();
+			if (!in || in->seq == _seq_expect-1) {
+				// No new data, sleep
+				std::this_thread::sleep_for(std::chrono::milliseconds{1});
+				return skip_option::skip_and_yield;
+			} else {
+				if (in->seq != _seq_expect) {
+					_stat_missed = in->seq - _seq_expect;
+				} else {
+					_stat_missed = 0;
+				}
+				_stat_processed++;
+				_seq_expect = in->seq+1;
+				return skip_option::run;
+			}
+		}
+
+		void _p_one_iteration() override {
```

Comment:
It is supposed to be synchronous but it is currently implemented this way because until SB 2.0 gets merged in, we don't have parallel synchronous callbacks.